api/keyconjurer/serverless_functions.go

```diff
@@ -1,130 +1,232 @@
-
-// Event holds incoming data from the user
-//  ShouldEncryptCreds determines if the user receieves encrypted credentials in their response
-type GetUserDataEvent struct {
-	Username           string `json:"username"`
-	Password           string `json:"password"`
-	Client             string `json:"client"`
-	ClientVersion      string `json:"clientVersion"`
-	ShouldEncryptCreds bool   `json:"shouldEncryptCreds"`
+type Handler struct {
+	crypt                   core.Crypto
+	cfg                     *settings.Settings
+	aws                     *aws.Provider
+	authenticationProviders providerMap
 }
 
-// GetUserDataEventHandler authenticates the user against OneLogin and retrieves a list of AWS application the user has available
-func GetUserDataEventHandler(event GetUserDataEvent) (*Response, error) {
-	logger := log.NewLogger(event.Client, event.ClientVersion, logrus.DebugLevel)
-	keyConjurerSettings, err := settings.NewSettings(logger)
+func NewHandler(cfg *settings.Settings) Handler {
+	client, err := aws.NewProvider(cfg.AwsRegion)
 	if err != nil {
-		return CreateResponseError(err.Error()), err
+		// TODO Probably shouldn't be a panic
+		panic(err)
 	}
 
-	auth, err := newAuthenticator(logger, keyConjurerSettings)
-	if err != nil {
-		return CreateResponseError(err.Error()), err
+	mfa := duo.New()
+	var prov core.CryptoProvider = &core.PassThroughProvider{}
+	if cfg.AwsKMSKeyID != "" {
+		prov = core.NewKMSProvider(&core.KMSProviderConfig{
+			KMSKeyID: cfg.AwsKMSKeyID,
+			// TODO: I think this might cause issues if the KMS secret is not in the same region as the Lambda
+			Session: session.New(),
+		})
 	}
 
-	// make new keyconjurer instance
-	client, err := NewKeyConjurer(auth, logger, keyConjurerSettings)
-	if err != nil {
-		return CreateResponseError(err.Error()), err
+	return Handler{
+		crypt: core.NewCrypto(prov),
+		cfg:   cfg,
+		aws:   client,
+		authenticationProviders: providerMap{
+			AuthenticationProviderOkta:     okta.Must(cfg.OktaHost, cfg.OktaToken, mfa),
+			AuthenticationProviderOneLogin: onelogin.New(cfg, mfa),
+		},
 	}
+}
 
-	// get username:password and decrypt if necessary
+// ClientProperties is information provided by the client about itself.
+//
+// This should not be relied on existing as it is user-provided information.
+// Newer versions of KeyConjurer place this information in the User-Agent header of their requests.
+// Older versions send it in their POST bodies in the GetUserData and GetAwsCreds endpoints.
+type ClientProperties struct {
+	Name    string `json:"client"`
+	Version string `json:"clientVersion"`
+}
 
-	user, err := client.providerClient.GetUserCredentials(event.Username, event.Password)
-	if err != nil {
-		return CreateResponseError(err.Error()), errors.New("unable to get user information with current provider")
+// FromRequestHeader updates the current properties from the given request's headers
+func (c *ClientProperties) FromRequestHeader(r *http.Request) bool {
```

Comment:
client version information is communicated in the user agent header rather than in the client body. this is done because some endpoints (like list providers) are technically GET requests and so cannot have a post body